core/modules/replica/DatabaseMySQL.cc

```diff
@@ -664,4 +691,110 @@ void Connection::_assertTransaction(bool inTransaction) const {
     }
 }
 
+
+ConnectionPool::Ptr ConnectionPool::create(ConnectionParams const& params,
+                                           size_t maxConnections) {
+    return Ptr(new ConnectionPool(params, maxConnections));
+}
+
+
+ConnectionPool::ConnectionPool(ConnectionParams const& params, size_t maxConnections)
+    :   _params(params),
+        _maxConnections(maxConnections) {
+}
+
+
+Connection::Ptr ConnectionPool::allocate() {
+
+    string const context = "ConnectionPool::" + string(__func__) + "  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    unique_lock<mutex> lock(_mtx);
+
+    // Open a new connection and return it right away if the limit hasn't
+    // been reached yet.
+    //
+    // TODO: the factory method called below may put a calling thread
+    // in the blocking state while the (database) service becomes available.
+    // This will prevent operations with the pool by other threads. Investigate
+    // a non-blocking algorithm.
+
+    if (_availableConnections.size() + _usedConnections.size() < _maxConnections) {
+        auto conn = Connection::open(_params);
+        _usedConnections.push_back(conn);
+        return conn;
+    }
+
+    // Otherwise grab an existing one (which may require to wait before
+    // it would become available).
+
+    auto const self = shared_from_this();
+    _available.wait(lock, [self]() {
+        return not self->_availableConnections.empty();
+    });
+    Connection::Ptr conn = _availableConnections.front();
+    _availableConnections.pop_front();
+    _usedConnections.push_back(conn);
+
+    return conn;
+}
+
+
+void ConnectionPool::release(Connection::Ptr const& conn) {
+
+    string const context = "ConnectionPool::" + string(__func__) + "  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    unique_lock<mutex> lock(_mtx);
+
+    // Move it between queues.
+
+    size_t numRemoved = 0;
+    _usedConnections.remove_if(
```

Comment:
`remove_if` is C++20 method, not sure we are allowed to use it.

Reply:
That's certainly true for the algorithm `std::remove_if`. However, what I'm using here is a method of class `std::list::remove_if` which is available in the Standard library before `C++20`. See:
https://en.cppreference.com/w/cpp/container/list/remove
Unlike the newer version that pre-version-20 method won't return any result. And I don't expect it here.

Reply:
OK, my fault, I looked at the wrong page.